src/js/quick-hull-3d/face.js

```diff
@@ -0,0 +1,271 @@
+var Face = function() {
+    this.mark = this.constructor.VISIBLE;
+};
+
+Face.VISIBLE = 1;
+Face.NON_CONVEX = 2;
+Face.DELETED = 3;
+
+Face.prototype.computeCentroid = function() {
+    var currentEdge = this.halfEdge0;
+    this.centroid = List.zerovector(CSNumber.real(3));
+
+    do {
+        this.centroid = List.add(this.centroid, currentEdge.head.point);
+        currentEdge = currentEdge.next;
+    } while (currentEdge !== this.halfEdge0);
+
+    this.centroid = List.scaldiv(this.numberOfVertices, this.centroid);
+};
+
+Face.prototype.computeNormal = function(minArea) {
+    var halfEdgeMax, lenSqrMax, lenMax,
+        currentEdge, lenSqr,
+        headPoint, tailPoint, u, dot;
+
+    var halfEdge1 = this.halfEdge0.next;
+    var halfEdge2 = halfEdge1.next;
+
+    var point0 = this.halfEdge0.head.point;
+    var point2 = halfEdge1.head.point;
+
+    var d2p = List.sub(point0, point2);
+    var d1p;
+
+    this.normal = List.zerovector(CSNumber.real(3));
+    this.numberOfVertices = 2;
+
+    while (halfEdge2 !== this.halfEdge0) {
+        d1p = List.turnIntoCSList(d2p.value.slice());
+
+        point2 = halfEdge2.head.point;
+        d2p = List.sub(point0, point2);
+
+        this.normal = List.add(this.normal, List.cross(d1p, d2p));
+
+        halfEdge2 = halfEdge2.next;
+        this.numberOfVertices++;
+    }
+
+    this.area = List.abs(this.normal);
+    this.normal = List.scaldiv(this.area, this.normal);
+
+    if (minArea === void 0) {
+        return;
+    }
+
+    if (this.area < minArea) {
+        halfEdgeMax = null;
+        lenSqrMax = 0;
+
+        currentEdge = this.halfEdge0;
+
+        do {
+            lenSqr = currentEdge.lengthSqr();
+
+            if (lenSqr > lenSqrMax) {
+                halfEdgeMax = currentEdge;
+                lenSqrMax = lenSqr;
+            }
+
+            currentEdge = currentEdge.next;
+        } while (currentEdge !== this.halfEdge0);
+
+        headPoint = halfEdgeMax.head.point;
+        tailPoint = halfEdgeMax.tail().point;
+
+        lenMax = Math.sqrt(lenSqrMax);
+
+        u = List.scaldiv(lenMax, List.sub(headPoint, tailPoint));
```

Comment:
I think this is wrong, since List.scaldiv expects the scalar to be a CSNumber. CSNumber.real(lenSqrMax) could help.